### PR TITLE
Add strings using Binary addition

### PR DESCRIPTION
**Pull request by Hiba-alkurd**

### CODE REVIEW

File: AddBinary/main.cs [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+            bool isZero = true;
+            while (isZero && result.Count > 1)
+            {
+                if (result.Peek() == '0') result.Pop();
+                else isZero = false;
+            }
+            return new string(result.ToArray());
+        }
+
+        public (char, char) FindSum(char a, char b, char carry)
```
Comment (ziadadeela) [inline AddBinary/main.cs]: It's better to name it something related to binary summation so that the consumer knows how to use it. For example, FindBinarySum

File: AddBinary/main.cs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        {
+            int sum = a + b + carry - '0'*3;
+            if (sum == 0) return ('0', '0');
+            if (sum == 1) return ('1', '0');
+            if (sum == 2) return ('0', '1');
+            if (sum == 3) return ('1', '1');
+            return ('0', '0');
```
Comment (ziadadeela) [inline AddBinary/main.cs]: I would say throwing an exception is better than returning a wrong default value. what do you think?

File: AddBinary/main.cs [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    public class main
+    {
+        public string AddBinary(string a, string b)
+        {
+            Stack<char> result = new Stack<char>();
+            
+            int lengthDiff = a.Length > b.Length ? a.Length - b.Length : b.Length - a.Length;
+            string appendZeros = new string('0',lengthDiff);
+            
+            if (a.Length > b.Length) b = appendZeros + b;
```
Comment (ziadadeela) [inline AddBinary/main.cs]: sometimes we may need the original value of the passed param. So in this case, I would define a new variable instead of changing the value of b directly. This will also give us an idea about the new value we assigned.

File: AddBinary/main.cs [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+            while (isZero && result.Count > 1)
+            {
+                if (result.Peek() == '0') result.Pop();
+                else isZero = false;
+            }
+            return new string(result.ToArray());
+        }
+
+        public (char, char) FindSum(char a, char b, char carry)
+        {
```
Comment (ziadadeela) [inline AddBinary/main.cs:47]: you can add some validation around the passed values. a,b, carry should be either 0 or 1

File: AddBinary/main.cs (hunk 3 above)
Comment (ziadadeela) [reply AddBinary/main.cs]: After checking the code, can't we get rid of appending zeros? since it only adds extra iterations that are not needed.

File: AddBinary/main.cs [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+            char carry = '0';
+            for ( var i = a.Length - 1; i >= 0; i--)
+            {
+                (sum, carry) = FindSum(a[i], b[i], carry);
+                result.Push(sum);
+            }
+            result.Push(carry);
+
+            bool isZero = true;
+            while (isZero && result.Count > 1)
```
Comment (ziadadeela) [inline AddBinary/main.cs:38]: why is this needed ?

**Review by ziadadeela — CHANGES_REQUESTED** [review]

File: AddBinary/main.cs (hunk 2 above)
Comment (Hiba-alkurd) [reply AddBinary/main.cs]: using default values instead of exceptions prevents the program from crashing.
but exceptions are better to show invalid values and errors

File: AddBinary/main.cs (hunk 3 above)
Comment (Hiba-alkurd) [reply AddBinary/main.cs]: I'll see what i can do , 
it can be substituted with a loop at the end and extra condition in the first loop to check when a string is done

File: AddBinary/main.cs (hunk 5 above)
Comment (Hiba-alkurd) [reply AddBinary/main.cs:38]: to remove redundant zeroes at the end

**Review by ziadadeela — APPROVED** [review]
